Added internal virtual on TokenHandler

brentschmaltz opened the pull request:
Add support to AadIssuerValidator for ValidationParameters.
Removed the parameter BaseConfiguration from delegates as ValidationParameters has ConfigurationManager which should be used to obtain BaseConfiguration when needed.

ValidationParameters.PropertyBag was not instantiated.
Create ValidationParameters.PropertyBag and InstancePropertyBag on demand.
Copy logic from existing AadIssuerValidator, keeping as code the same as much as possible.
Copy logic from existing AadTokenValidationParametersExtension to AadValidationParametersExtension keeping as code the same as much as possible.
This copied code will need to be refactored into common base code.
Move CloudInstanceNameKey to constants file.

Added tests to AadTokenValidationParametersExtensionTests and MicrosoftIdentityIssuerValidatorTest that cover existing tests.

jmprieur commented on test/Microsoft.IdentityModel.TestUtils/TestUtilities.cs:
```diff
@@ -35,6 +35,78 @@ public class GetSetContext
 
     public static class TestUtilities
     {
+        internal static ValidationParameters CreateFromTokenValidationParameters(TokenValidationParameters tokenValidationParameters)
```
Nice!
Do we want to do something for the delegates too?

jmprieur replied:
Given the number of common fields, should there be a common base class with these?

brentschmaltz replied:
@jmprieur perhaps in the future, we aren't there right now.

keegan-caruso commented on src/Microsoft.IdentityModel.Validators/AadIssuerValidator/AadIssuerValidator.Internal.cs:
```diff
@@ -0,0 +1,141 @@
+        /// <param name="issuer">Issuer to validate (will be tenanted).</param>
+        /// <param name="securityToken">Received security token.</param>
+        /// <param name="validationParameters">The <see cref="ValidationParameters"/> to be used for validating the token.</param>
+        /// <param name="callContext"></param>
```
A few these are not filled out

keegan-caruso commented on src/Microsoft.IdentityModel.Validators/AadIssuerValidator/AadIssuerValidator.Internal.cs:
```diff
@@ -0,0 +1,141 @@
+                var issuerVersion = GetTokenIssuerVersion(securityToken);
+                var effectiveConfigurationManager = GetEffectiveConfigurationManager(issuerVersion);
+
+                string aadIssuer = null;
```
```suggestion
                string aadIssuer;
```

This gets unconditionally wrote to below.

keegan-caruso commented on src/Microsoft.IdentityModel.Validators/AadValidationParametersExtension.Internal.cs:
```diff
@@ -0,0 +1,254 @@
+
+            IssuerSigningKeyValidationDelegate issuerSigningKeyValidationDelegate1 = (securityKey, securityToken, validationParameters, callContext) =>
+            {
+                // TODO remove GetAwaiter
```
Why can't we do this now and make the method async? Better to do this sooner I would think

brentschmaltz replied:
The delegate is not async, so i think we have to add some additional work to make this happen.

jmprieur commented on src/Microsoft.IdentityModel.Tokens/Validation/ValidationParameters.cs:
```diff
@@ -374,7 +380,9 @@ public string NameClaimType
         /// This allows addition of parameters that could be used in custom token validation scenarios.
         /// </summary>
-        public IDictionary<string, object> PropertyBag { get; }
+        public IDictionary<string, object> PropertyBag => _propertyBag ??
```
and this one is not unique to the instance? is it shared between ValidationParameters that were cloned from each other?

brentschmaltz replied:
Yes shared between all instances.

jmprieur commented on src/Microsoft.IdentityModel.Validators/AadIssuerValidator/AadIssuerValidator.cs:
```diff
@@ -19,7 +19,7 @@ namespace Microsoft.IdentityModel.Validators
     /// <summary>
     /// Generic class that validates the issuer for either JsonWebTokens or JwtSecurityTokens issued from the Microsoft identity platform (AAD).
     /// </summary>
-    public class AadIssuerValidator
```
How much copy/paste is there in the partial implementation vs how much can we have in common? I understand that the input (ValidationParameters, context) will be different, and that the output is an Error, not an exception thrown, but can the logic be a bit more common?

brentschmaltz replied:
Certainly, that can be done. Not the goal right now. The goal is to enable upper layers to make the calls.

jmprieur commented on src/Microsoft.IdentityModel.Validators/AadValidationParametersExtension.Internal.cs:
```diff
@@ -0,0 +1,254 @@
+                // TODO remove GetAwaiter
+                BaseConfiguration configuration = null;
+                if (validationParameters.ConfigurationManager != null)
+                    configuration = validationParameters.ConfigurationManager.GetBaseConfigurationAsync(System.Threading.CancellationToken.None).ConfigureAwait(false).GetAwaiter().GetResult();
```
Is it ok to have a sync over async here? I guess we need to configuration and need to block?

jmprieur commented on src/Microsoft.IdentityModel.Validators/AadValidationParametersExtension.Internal.cs:
```diff
@@ -0,0 +1,254 @@
+        /// <param name="securityToken">The <see cref="SecurityToken"/> being validated, could be a JwtSecurityToken or JsonWebToken.</param>
+        /// <param name="configuration">The <see cref="BaseConfiguration"/> provided.</param>
+        /// <returns><c>true</c> if the issuer of the signing key is valid; otherwise, <c>false</c>.</returns>
+        internal static bool ValidateIssuerSigningKey(SecurityKey securityKey, SecurityToken securityToken, BaseConfiguration configuration)
```
This method is common with AadTokenValidationParametersExtensions (to the letter)

jmprieur commented on src/Microsoft.IdentityModel.Validators/AadValidationParametersExtension.Internal.cs:
```diff
@@ -0,0 +1,254 @@
+
+                    if (!string.Equals(signingKeyCloudInstanceName, configurationCloudInstanceName, StringComparison.Ordinal))
+                        throw LogHelper.LogExceptionMessage(
+                            new SecurityTokenInvalidCloudInstanceException(
```
this is the only line that changes in this method vs in AadValidationParametersExtension

jmprieur commented on src/Microsoft.IdentityModel.Validators/AadValidationParametersExtension.Internal.cs:
```diff
@@ -0,0 +1,254 @@
+            }
+        }
+
+        private static JsonWebKey GetJsonWebKeyBySecurityKey(OpenIdConnectConfiguration configuration, SecurityKey securityKey)
```
Same method and content in both

jmprieur commented on src/Microsoft.IdentityModel.Validators/AadValidationParametersExtension.Internal.cs:
```diff
@@ -0,0 +1,254 @@
+            return null;
+        }
+
+        private static string GetTid(SecurityToken securityToken)
```
Same method and content in both

jmprieur commented on src/Microsoft.IdentityModel.Validators/AadValidationParametersExtension.Internal.cs:
```diff
@@ -0,0 +1,254 @@
+            }
+        }
+
+        private static void EnforceSingleClaimCaseInsensitive(IEnumerable<string> keys, string claimType)
```
Same method and content in both

jmprieur reviewed (commented):
I've asked a few questions.
A lot of duplicated code in AadTokenValidationParametersExtension / AadValidationParametersExtension.Internal

jmprieur commented on src/Microsoft.IdentityModel.Tokens/TokenHandler.Internal.cs:
```diff
@@ -0,0 +1,33 @@
+            CallContext callContext,
+            CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
```
throw new NotImplementedExeption("Make sure you override this method in the derived class");

jmprieur commented on src/Microsoft.IdentityModel.Tokens/TokenHandler.Internal.cs:
```diff
@@ -0,0 +1,33 @@
+            CallContext callContext,
+            CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
```
same?

brentschmaltz replied:
Addressed

jmprieur reviewed (approved):
LGTM
Thanks @brentschmaltz

HP712 committed:
Added internal virtual on TokenHandler
Add AadIssuer with ValidationParameters
Added tests to AadIssuerValidator for ValidationParameters
